### PR TITLE
BMS Charging Update (fixed)

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
## Summary

Relevant Jira: [ELEC-41](https://ubcformulaelectric.atlassian.net/browse/ELEC-41)
#### Init State:
- If charger is connected, the board waits for a CAN message to enable charging (Enters Pre-Charge)
#### Pre-Charge State:
- Failing Pre-Charge for charging will disable the CAN message . It will also send it back to Init State if Pre-Charge has not failed more than three times and if it is has been more than three times then it is sent into Fault State.
#### All States:
- If the charger is connected, it checks if the charger has thrown a fault it will get fault state.
- If the charger is connected and AIRs- opens (Shutdown loop activates), sends BMS to fault state.
- If the charger is not connected it checks if the car is in charging state which would indicate that the charger was disconnected midway through charging. It would then go into fault state.
#### Charge State:
- Disabling CAN message for charging and charger has been put into RunOnExit.  
- Checks all at once if the CAN message is disabled, external shutdown loop has triggered or if charging has completed and enters Init State if any are true.
- Checking if charging is complete is done by comparing the current being supplied to 0.05C charging.

## Testing Done

- With no enabling CAN message to charge while the charger is connected in Init state, the BMS should remain in Init state waiting for charging CAN message.
- When the charger is connected in Init state and the CAN message is given, the BMS should go into Precharge state
- After charger is connected and the CAN message enabling charging is sent and the system waits for slightly longer, BMS should go into Charge state (Successful precharge should close AIRs+.
- Starting in Charge state, with CAN message enabled and the charger connected with less than full charge, the BMS will remain in Charge state.
- In Charge state, when current from charger indicates full charge on the cells, the BMS stops charging, switches off the charging CAN message and goes back into Init State and opens AIRs+.
- In Charge state, will exit back to Init state in similar manner when false CAN charging message is sent.
- In Charge state, will go into fault state if the charger gives a fault. Will also disable CAN charging message.
- In Charge state, will go into fault state if charger is connected and the shutdown loop activates (AIRs- is open)
- Modified earlier test with charger disconnecting while in Charge state sending BMS to fault state.

## Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

[ELEC-41]: https://ubcformulaelectric.atlassian.net/browse/ELEC-41?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: boards/BMS/Inc/App/states/App_ChargeState.h [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
 #include "App_SharedStateMachine.h"
-
+#include "App_Accumulator.h"
 /**
  * Get a pointer to the Charge State.
  * @return A pointer to the Charge State. THIS SHOULD NOT BE MODIFIED.
  */
 const struct State *App_GetChargeState(void);
+// 0.05C is standard for a boundary to consider full charge
+#define CURRENT_AT_MAX_CHARGE (0.05f * C_RATE_TO_AMPS);
```
Comment (will-chaba) [inline boards/BMS/Inc/App/states/App_ChargeState.h:11]: This should be placed above function definitions, also add an empty line to end of file

File: boards/BMS/Test/Src/Test_Bms.cpp [hunk 2]
```diff
@@ -1,3 +1,72 @@
 #include "Test_Bms.h"
 
 DEFINE_FFF_GLOBALS;
+//
```
Comment (will-chaba) [inline boards/BMS/Test/Src/Test_Bms.cpp:4]: Was the addition of this file intentional? Looks like it could be removed

File: can-bus/json/BMS/BMS_rx.json [hunk 3]
```diff
@@ -3,6 +3,7 @@
         "DCM_Vitals",
         "FSM_Vitals",
         "PDM_Vitals",
-        "DIM_Vitals"
+        "DIM_Vitals",
+        "DEBUG_ChargingSwitch"
     ]
 }
```
Comment (will-chaba) [inline can-bus/json/BMS/BMS_rx.json:9]: missing newline at end of file

File: can-bus/json/DEBUG/DEBUG_alerts.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+    "msg_id": 700,
+    "alerts": []
+}
```
Comment (will-chaba) [inline can-bus/json/DEBUG/DEBUG_alerts.json:4]: missing newline at end of file

File: can-bus/json/DEBUG/DEBUG_enum.json [hunk 5]
```diff
@@ -0,0 +1,2 @@
+{
+}
```
Comment (will-chaba) [inline can-bus/json/DEBUG/DEBUG_enum.json:2]: missing newline at end of file

File: can-bus/json/DEBUG/DEBUG_rx.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "messages": []
+}
```
Comment (will-chaba) [inline can-bus/json/DEBUG/DEBUG_rx.json:3]: missing newline at end of file

File: can-bus/json/DEBUG/DEBUG_tx.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+        "msg_id": 701,
+        "cycle_time": 100,
+        "description": "Value set by user to start charging BMS.",
+        "signals": {
+            "StartCharging": {
+                "bits": 1
+            }
+        }
+    }
+}
```
Comment (will-chaba) [inline can-bus/json/DEBUG/DEBUG_tx.json:12]: missing newline at end of file

**Review by will-chaba — CHANGES_REQUESTED** [review]
Looks good, just some minor items to clean up

**Review by will-chaba — DISMISSED** [review]
Looks good to me

File: boards/BMS/Inc/App/states/App_ChargeState.h [hunk 8]
```diff
@@ -1,9 +1,11 @@
 #pragma once
 
 #include "App_SharedStateMachine.h"
-
+#include "App_Accumulator.h"
+// 0.05C is standard for a boundary to consider full charge
+#define CURRENT_AT_MAX_CHARGE (0.05f * C_RATE_TO_AMPS);
```
Comment (gtaharaedmonds) [inline boards/BMS/Inc/App/states/App_ChargeState.h:6]: This isn't used in any other file, so you should define it in the .c file instead to limit scope

File: Pipfile.lock
Comment (gtaharaedmonds) [inline Pipfile.lock:1]: Delete this file, we don't track this anymore

File: boards/BMS/Src/App/states/App_InitState.c [hunk 9]
```diff
@@ -30,24 +32,32 @@ static void InitStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+            // to continue into Pre-Charge State, then Charge State
+            if (is_charger_connected && App_CanRx_DEBUG_ChargingSwitch_StartCharging_Get())
+            {
+                App_SharedStateMachine_SetNextState(state_machine, App_GetPreChargeState());
+            }
+            // If charger isn't connected, go into Pre-Charge State, then Drive State
+            else if (!is_charger_connected)
+            {
+                App_SharedStateMachine_SetNextState(state_machine, App_GetPreChargeState());
+            }
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_InitState.c:58]: I feel like the use of an if/else here is confusing since both conditions lead to precharge state. 
Consider changing to a single if statement, something like:
```c
bool precharge_for_charging = is_charger_connected && App_CanRx_DEBUG_ChargingSwitch_StartCharging_Get();
if(precharge_for_charging || !is_charger_connected)
{
    App_SharedStateMachine_SetNextState(state_machine, App_GetPreChargeState());
}
```
```

File: boards/BMS/Src/App/states/App_InitState.c [hunk 10]
```diff
@@ -30,24 +32,32 @@ static void InitStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
         }
 #endif
-        //      #else?
-        struct HeartbeatMonitor *hb_monitor    = App_BmsWorld_GetHeartbeatMonitor(world);
-        const bool               is_missing_hb = !App_SharedHeartbeatMonitor_Tick(hb_monitor);
-        App_CanTx_BMS_Warnings_MissingHeartBeat_Set(is_missing_hb);
-        if (is_missing_hb)
-        {
-            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
-        }
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_InitState.c:50]: You should revert this change here, I don't think this is supposed to be part of your PR

File: boards/BMS/Src/App/states/App_PreChargeState.c [hunk 11]
```diff
@@ -62,20 +62,24 @@ static void PreChargeStateRunOnTick100Hz(struct StateMachine *const state_machin
...
+            if (next_state == App_GetFaultState())
+            {
+                App_PrechargeRelay_ResetFaultCounterVal(precharge_relay);
+            }
             App_SharedStateMachine_SetNextState(state_machine, next_state);
         }
-        else if (missing_hb)
-        {
-            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
-        }
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_PreChargeState.c:78]: Did you mean to delete all this heartbeat stuff?

File: boards/BMS/Src/App/states/App_PreChargeState.c [hunk 12]
```diff
@@ -62,20 +62,24 @@ static void PreChargeStateRunOnTick100Hz(struct StateMachine *const state_machin
...
         if (has_precharge_fault)
         {
             const struct State *next_state =
                 (precharge_fault_limit_exceeded) ? App_GetFaultState() : App_GetInitState();
+            App_CanRx_DEBUG_ChargingSwitch_StartCharging_Update(false);
+            if (next_state == App_GetFaultState())
+            {
+                App_PrechargeRelay_ResetFaultCounterVal(precharge_relay);
+            }
             App_SharedStateMachine_SetNextState(state_machine, next_state);
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_PreChargeState.c:78]: This logic might be clearer like:
```c
if(precharge_fault_limit_exceeded)
{
    App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
    App_PrechargeRelay_ResetFaultCounterVal(precharge_relay);
}
else
{
    App_SharedStateMachine_SetNextState(state_machine, App_GetInitState());
}
```

I feel like setting the next state ptr and then `if`ing based on that is confusing

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 13]
```diff
@@ -43,24 +44,16 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
 
-        uint8_t    segment = 0U;
-        uint8_t    cell    = 0U;
-        const bool has_reached_max_v =
-            App_Accumulator_GetMaxVoltage(accumulator, &segment, &cell) > MAX_CELL_VOLTAGE_THRESHOLD;
+        const bool charging_completed = App_TractiveSystem_GetCurrent(ts) <= CURRENT_AT_MAX_CHARGE;
 
-        App_CanTx_BMS_Faults_ChargerDisconnectedInChargeState_Set(is_charger_disconnected);
         App_CanTx_BMS_Faults_ChargerFault_Set(has_charger_faulted);
-        App_CanTx_BMS_Charger_IsChargingComplete_Set(has_reached_max_v);
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_ChargeState.c:53]: IsChargingComplete and ChargerDisconnectedInChargeState are important signals - you should change this back so they are actually being set

File: scripts/codegen/CAN/App_CanMsgs.dbc
Comment (gtaharaedmonds) [inline scripts/codegen/CAN/App_CanMsgs.dbc:1]: We don't track the DBC file anymore, you should delete this

File: boards/BMS/Src/App/states/App_AllStates.c [hunk 14]
```diff
@@ -155,6 +156,23 @@ bool App_AllStatesRunOnTick100Hz(struct StateMachine *const state_machine)
...
+            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+        }
+    }
+    else if (App_SharedStateMachine_GetCurrentState(state_machine) == App_GetChargeState())
+    {
+        status = false;
+        App_CanTx_BMS_Faults_ChargerDisconnectedInChargeState_Set(true);
+        App_CanRx_DEBUG_ChargingSwitch_StartCharging_Update(false);
+        App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+    }
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_AllStates.c:175]: Why is this check here now vs. in App_ChargeState previously? I think it makes more sense there since that's where all the other charging faults are handled

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 13 above)
Comment (gtaharaedmonds) [reply boards/BMS/Src/App/states/App_ChargeState.c:53]: Also the faults here don't actually put the board into fault state

**Review by gtaharaedmonds — COMMENTED** [review]
Overall looks great but I think needs some work, take a look at the comments.
I haven't reviewed tests yet, will do so when other stuff is ironed out.

File: boards/BMS/Src/App/states/App_AllStates.c (hunk 14 above)
Comment (will-chaba) [reply boards/BMS/Src/App/states/App_AllStates.c]: This was part of Edwin's charging diagram, it was found to not work here at all because of how the charger sends its fault signal, basically, it sends the fault signal whenever it is not currently charging, even if there is nothing "wrong", so checking in all-states will perpetually send the BMS into fault state. 

TLDR: moved the check back to charge state

File: scripts/codegen/CAN/App_CanMsgs.dbc
Comment (will-chaba) [reply scripts/codegen/CAN/App_CanMsgs.dbc:1]: what's the difference between this file and can-bus/dbcs/CanMsgs.dbc?

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 15]
```diff
@@ -25,48 +29,77 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        if (ignore_chgr_fault_counter >= CYCLES_TO_IGNORE_CHGR_FAULT)
         {
-            ignore_chgr_fault_counter++;
+            has_charger_faulted = App_Charger_HasFaulted(charger);
+            timeout_passed      = true;
+            UNUSED(ts);
+            // TODO: Re-implement charging_completed
+            //            if(-App_TractiveSystem_GetCurrent(ts) <=  0.85f && (App_TractiveSystem_GetCurrent(ts) < 0)){
+            //                charging_completed = true;
+            //            }
```
Comment (will-chaba) [inline boards/BMS/Src/App/states/App_ChargeState.c:54]: charging_completed was magically going true immediately even though the current was much greater than 0.85, could be a current sense glitch, but CAN read the value as significantly higher than that (greater than 2.0A minimum during our tests if I recall correctly.) The current plan is just to babysit the charger and manually stop charging using the CAN message, but this should absolutely be looked into further in the future time permitting

**Comment by will-chaba** [comment]
@gtaharaedmonds I'd request your review but you're the PR author

File: boards/BMS/Src/App/App_Accumulator.c [hunk 16]
```diff
@@ -306,7 +306,7 @@ bool App_Accumulator_CheckFaults(struct Accumulator *const accumulator, struct T
     float min_allowable_cell_temp = MIN_CELL_DISCHARGE_TEMP_DEGC;
 
     // if we are charging, max cell temp is 45C not 60C
-    if (App_TractiveSystem_GetCurrent(ts) < -3.0f)
+    if (App_TractiveSystem_GetCurrent(ts) > 3.0f)
```
Comment (KelmLelm) [inline boards/BMS/Src/App/App_Accumulator.c:309]: I'm not completely sure why the voltage we were checking for on the tractive system before was negative. Is it at all related to getting the ts voltage for charging finished?

File: boards/BMS/Src/App/App_TractiveSystem.c [hunk 17]
```diff
@@ -48,7 +48,7 @@ float App_TractiveSystem_GetCurrent(struct TractiveSystem *ts)
     float low_res_current  = App_TractiveSystem_GetLowResCurrent(ts);
     float high_res_current = App_TractiveSystem_GetHighResCurrent(ts);
 
-    if (low_res_current < HIGH_RES_MAX_CURRENT_READING)
+    if (low_res_current < HIGH_RES_MAX_CURRENT_READING && low_res_current > -HIGH_RES_MAX_CURRENT_READING)
```
Comment (KelmLelm) [inline boards/BMS/Src/App/App_TractiveSystem.c:51]: I'm not too sure what's going on here. Under different circumstances, can the ts return negative values for current?

File: boards/BMS/Src/App/states/App_AllStates.c [hunk 18]
```diff
@@ -140,19 +145,20 @@ bool App_AllStatesRunOnTick100Hz(struct StateMachine *const state_machine)
...
     App_CanTx_BMS_OkStatuses_BmsOk_Set(App_OkStatus_IsEnabled(bms_ok));
     App_CanTx_BMS_OkStatuses_ImdOk_Set(App_OkStatus_IsEnabled(imd_ok));
     App_CanTx_BMS_OkStatuses_BspdOk_Set(App_OkStatus_IsEnabled(bspd_ok));
 
-    const bool dcm_fault              = App_CanAlerts_BoardHasFault(DCM_ALERT_BOARD);
-    const bool fsm_fault              = App_CanAlerts_BoardHasFault(FSM_ALERT_BOARD);
-    const bool pdm_fault              = App_CanAlerts_BoardHasFault(PDM_ALERT_BOARD);
-    const bool dim_fault              = App_CanAlerts_BoardHasFault(DIM_ALERT_BOARD);
-    const bool fault_from_other_board = dcm_fault || fsm_fault || pdm_fault || dim_fault;
+    const bool dcm_fault = App_CanAlerts_BoardHasFault(DCM_ALERT_BOARD);
```
Comment (KelmLelm) [inline boards/BMS/Src/App/states/App_AllStates.c:156]: Are there any faults we can have while charging that would be dangerous if not detected from other boards?

File: boards/BMS/Src/App/App_Accumulator.c (hunk 16 above)
Comment (will-chaba) [reply boards/BMS/Src/App/App_Accumulator.c:309]: The TS current convention we want to use is that negative current is current leaving the pack (discharge), so charging current should be positive. I believe this was backwards before so that's why it had to switch

File: boards/BMS/Src/App/App_TractiveSystem.c (hunk 17 above)
Comment (will-chaba) [reply boards/BMS/Src/App/App_TractiveSystem.c:51]: Yes. Depending on charging vs discharging

File: boards/BMS/Src/App/states/App_AllStates.c (hunk 18 above)
Comment (will-chaba) [reply boards/BMS/Src/App/states/App_AllStates.c:156]: Charging is a process that is handled solely by the BMS, no other boards are involved, so there shouldn't be any need to check for faults from other boards while charging specifically. (additionally charging is often done outside the car where the other boards aren't connected) Hence why I added a conditional operator on `fault_from_other_board` on line 161

@gtaharaedmonds did I get that right?

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 15 above)
Comment (KelmLelm) [reply boards/BMS/Src/App/states/App_ChargeState.c:54]: Not sure if you changed the code from what's in the commented out block, but if the ts current is positive when charging, the negative sign in front of the App_Tract... would be incorrect. But if you have the check for if ts current is negative as well, then the charging_completed = true should never be reached so I'm not sure.

Also, unrelated I realized we didn't use the CURRENT_AT_MAX_CHARGE variable here and just put 0.85f.

File: boards/BMS/Src/App/states/App_AllStates.c (hunk 18 above)
Comment (KelmLelm) [reply boards/BMS/Src/App/states/App_AllStates.c:156]: That does make sense, if that's the case then this file should be good. I'll wait for some confirmation though

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 15 above)
Comment (will-chaba) [reply boards/BMS/Src/App/states/App_ChargeState.c:54]: Yeah the commented code was altered during testing, the reason we weren’t using the macro was a troubleshooting step, I’ll clean it up.

As for the negative sign in front of ts_current, it was reading as positive when it should have been negative, because the previous convention was to have power coming in be positive and it hasn’t fully been switched over yet. Again, I’ll clean up the commented block to be as you left it

**Review by will-chaba — APPROVED** [review]
Worked on car, battery no go boom :fire:

File: boards/BMS/Src/App/states/App_AllStates.c (hunk 18 above)
Comment (will-chaba) [reply boards/BMS/Src/App/states/App_AllStates.c:156]: I confirmed with @Eddy13013 that other boards will always be disconnected when charging

**Comment by KelmLelm** [comment]
> Worked on car, battery no go boom 🔥

SICKK
